python/ray/workflow/step_executor.py

```diff
@@ -197,7 +229,13 @@ def commit_step(store: workflow_storage.WorkflowStorage,
     """
     from ray.workflow.common import Workflow
     if isinstance(ret, Workflow):
-        store.save_subworkflow(ret)
+        assert not ret.executed
```

Comment:
I don't quite like the new logic here, but I don't like the old logic as well 😅
The new one use a lot of apis inside workflow_storage which originally we want to hide from up layer.

The old one coupled storage with app layer a lot.

We probably can think about the positioning of `workflow_storage` a little bit.

Let's revisit what we have right now (with this PR):

* storage: an abstraction for database
* workflow_storage: contains the schema for workflow and use storage to write
  * key for different things
  * logic for how to updating things for different cases
* serialization: how to convert objects to bytes + dumps

I feel once we have open in workflow_storage, we'll decouple serialization/workflow_storage (right now, we need _key_ and also serialization.dump_to_storage() to make things work.

Then the next thing is about how to reduce the application-related code in workflow storage. One thing we can do is to move them to the place where they are used like what's doing here. 

So basically, we are going to have:

* storage: an abstraction for database
* workflow_storage: contains the schema for workflow build on top of the storage
* serialization: how to convert objects to bytes, like a pickler.
* application logic will be put in the application layer. and use serialization + workflow_storage to support specific functions